### PR TITLE
ARROW-15584: [C++] Add support for Substrait's RelCommon::Emit

### PR DESCRIPTION
**Pull request by vibhatha**
Adding emit feature for Substrait plan deserialization. 

This PR covers emits for `read`, `filter`, `project`, `join` and `aggregate` operations.

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @westonpace @jeroen please take a look.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 1]
```diff
@@ -248,19 +286,56 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+        expressions.emplace_back(des_expr);
+      }
+      while (!new_fields.empty()) {
+        auto field = new_fields.back();
+        ARROW_ASSIGN_OR_RAISE(
+            project_schema,
+            project_schema->AddField(
+                input.num_columns + static_cast<int>(new_fields.size()) - 1,
+                std::move(field)));
+        new_fields.pop_back();
```
Comment (jvanstraten) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:317]: I'm pretty sure this adds the fields in reverse order, for what it's worth. Why is this in a separate loop?

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 2]
```diff
@@ -399,17 +506,45 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+      FieldVector output_fields;
+      output_fields.reserve(key_field_ids.size() + agg_src_field_ids.size());
+      // extract aggregate fields to output schema
+      for (int id = 0; id < static_cast<int>(agg_src_field_ids.size()); id++) {
+        output_fields.emplace_back(input_schema->field(agg_src_field_ids[id]));
+      }
+      // extract key fields to output schema
+      for (int id = 0; id < static_cast<int>(key_field_ids.size()); id++) {
+        output_fields.emplace_back(input_schema->field(key_field_ids[id]));
+      }
```
Comment (jvanstraten) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:529]: Wrong order; keys come first.

> The list of distinct columns from each grouping set (ordered by their first appearance) followed by the list of measures in declaration order, [...]

https://substrait.io/relations/logical_relations/#aggregate-operation

File: cpp/src/arrow/engine/substrait/relation_internal.h [hunk 3]
```diff
@@ -38,6 +38,8 @@ struct DeclarationInfo {
 
   /// The number of columns returned by the declaration.
   int num_columns;
+
+  std::shared_ptr<Schema> output_schema;
```
Comment (jvanstraten) [inline cpp/src/arrow/engine/substrait/relation_internal.h:42]: I feel like this supersedes `num_columns`; `num_columns` should always equal the number of fields in the schema. I only put that there because I didn't want to have to bother with the whole schema yet.

**Review by jvanstraten — CHANGES_REQUESTED** [review]
I only checked for inconsistencies with Substrait, so not for C++ or Acero-related problems or code quality. Emit handling looks good to me from that perspective, but I did find a few schema deduction problems.

File: cpp/src/arrow/engine/substrait/relation_internal.h (hunk 3 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.h:42]: Yes this make sense. I will update this.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 4]
```diff
@@ -199,12 +217,22 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
-          num_columns};
+      if (HasEmit(read)) {
+        ARROW_ASSIGN_OR_RAISE(auto emit_expressions, GetEmitInfo(read, base_schema));
+        return DeclarationInfo{
+            compute::Declaration::Sequence(
+                {{"scan",
+                  dataset::ScanNodeOptions{std::move(ds), std::move(scan_options)}},
+                 {"project", compute::ProjectNodeOptions{std::move(emit_expressions)}}}),
+            static_cast<int>(emit_expressions.size()), std::move(base_schema)};
+      } else {
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:230]: @jvanstraten brought up a good point to me offline which is that we should only do this else clause if things are direct.  Also, I think `has_emit` is only available in newer protoc versions.  Can we change all these if/else blocks to be...

```
switch (rel.emit_kind_case()) {
  case substrait::RelCommon::EmitKindCase::kDirect:
    // return no-emit case
  case substrait::RelCommon::EmitKindCase::kEmit:
    // return emit-case
  default:
    // return invalid status
}
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 1 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc]: Thanks for noting this one. Redundant and I missed this.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 4 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:229]: There is a question about this, I think the `Rel` doesn't have the `emit_kind_case`, but the `RelCommon` does. 
So first we need to see if `has_common()` returns true and then check the `common().emit_kind_case()`. But if `has_common()` returns false we have to return the `no_emit` declaration. Is this a wrong interpretation of things?

Comment (westonpace) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:229]: That is a correct interpretation.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:516]: @jvanstraten I also noticed this, but I forget to leave a comment about it. This is probably a separate JIRA because of the order used in the `aggregate_node.cc`[1]. Please refer to the comment in this line and the two loops after that. The aggregate fields appened first and then the key fields. One thing we can do is swap the response here.

cc @westonpace 

[1]. https://github.com/apache/arrow/blob/50a7d15dfb4cbc4dd449ff2bb3ba2b1cde62a3ab/cpp/src/arrow/compute/exec/aggregate_node.cc#L345

Comment (jvanstraten) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:516]: Based on the comment this looks like intentional behavior of Arrow, so I don't think aggregate node is going to be adjusted to match Substrait. So that just means there should be a project node inserted behind the aggregate node that moves the columns around accordingly, right? I guess you could fix that in a separate JIRA/PR though. Maybe add a FIXME comment in that case?

Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:516]: Yes, the requirement in Acero may be static here. 
We can use the project to swap things around and document it properly. Probably we can do it in this PR as well. 

cc @westonpace

**Review by westonpace — COMMENTED** [review]
It seems I had left this review in the pending state.  Apologies.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 5]
```diff
@@ -121,6 +155,78 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
...
+    return Status::OK();
+  }
+
+  std::shared_ptr<Table> input_table;
+  std::string file_prefix;
+  std::unique_ptr<arrow::internal::TemporaryDir>& tempdir;
+  std::string data_file_path;
+};
+
+struct EmitValidate {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:181]: Done.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 6]
```diff
@@ -121,6 +155,78 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
...
+  EmitValidate(const std::shared_ptr<Schema> output_schema,
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:182]: Yes, you're right the wording is wrong. The intention was to use both the data generator plus this function to be used in end-to-end cases. So we can use it in other cases too.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 7]
```diff
@@ -1814,5 +1920,1049 @@ TEST(Substrait, AggregateBadPhase) {
...
+      [3, 5, 20],
+      [4, 1, 30],
+      [2, 1, 40],
+      [5, 5, 50],
+      [2, 2, 60]
+  ])"});
+
+  std::string file_prefix = "serde_project_test";
+  ASSERT_OK_AND_ASSIGN(auto tempdir,
+                       arrow::internal::TemporaryDir::Make("substrait_project_tempdir"));
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:1936]: I agree, but this is the case where we were actually testing the file system parts in Acero-Substrait. On the other hand, there is a pending JIRA to add more end-to-end tests then we could write these tests there. So the effort is not a waste. I can modify this to use NamedTable. That's a nice suggestion.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 8]
```diff
@@ -1814,5 +1920,1049 @@ TEST(Substrait, AggregateBadPhase) {
...
+                               field("C", int32()), field("equal", boolean())});
+  auto expected_table = TableFromJSON(output_schema, {R"([
+    [1, 1, 10, true],
+    [3, 5, 20, false],
+    [4, 1, 30, false],
+    [2, 1, 40, false],
+    [5, 5, 50, true],
+    [2, 2, 60, true]
+  ])"});
+  EmitValidate(std::move(output_schema), std::move(expected_table), exec_context, buf)();
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:2036]: Yes, I refactored this to `CheckRoundTripResult`.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 9]
```diff
@@ -121,6 +155,78 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
...
+               compute::ExecContext& exec_context, std::shared_ptr<Buffer>& buf,
+               const std::vector<int>& include_columns = {})
+      : output_schema(output_schema),
+        expected_table(expected_table),
+        exec_context(exec_context),
+        buf(buf),
+        include_columns(include_columns) {}
+  void operator()() {
+    for (auto sp_ext_id_reg :
+         {std::shared_ptr<ExtensionIdRegistry>(), MakeExtensionIdRegistry()}) {
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:193]: I think I just adopted what is being done in a previous test without paying that much attention to it. We could just use the `MakeExtensionRegistry`. I already changed it [ToProto-Read](https://github.com/apache/arrow/pull/13401). 

And I updated here too.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 10]
```diff
@@ -1814,5 +1920,1049 @@ TEST(Substrait, AggregateBadPhase) {
...
+  auto output_schema = schema({
+      field("A", int32()),
+      field("B", int32()),
+      field("C", int32()),
+      field("D", int32()),
+      field("X", int32()),
+      field("Y", int32()),
+      field("Z", int32()),
+      field("W", int32()),
+  });
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: Done.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 11]
```diff
@@ -1814,5 +1920,1049 @@ TEST(Substrait, AggregateBadPhase) {
...
+  ASSERT_OK_AND_ASSIGN(auto buf, internal::SubstraitFromJSON("Plan", substrait_json));
+  auto output_schema = schema({
+      field("A", int32()),
+      field("B", int32()),
+      field("C", int32()),
+      field("D", int32()),
+      field("__fragment_index_l", int32()),
+      field("__batch_index_l", int32()),
+      field("__last_in_fragment_l", boolean()),
+      field("__filename_l", utf8()),
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: Yes, I added `NamedTable` and it is pretty neat now.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:518]: On second thoughts, it would be better to solve this one in another PR. Because I am not quite sure if this would break R test cases.

**Comment by vibhatha** [comment]
@westonpace I updated the PR.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 12]
```diff
@@ -94,9 +135,11 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
         const substrait::ReadRel::NamedTable& named_table = read.named_table();
         std::vector<std::string> table_names(named_table.names().begin(),
                                              named_table.names().end());
-        ARROW_ASSIGN_OR_RAISE(compute::Declaration source_decl,
+        ARROW_ASSIGN_OR_RAISE(compute::Declaration no_emit_declaration,
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:138]: Can you revert this change?  From the perspective of `FromProto` I think this is still the "source_decl".  The name "no_emit_declaration" only really makes sense inside `ProcessEmit`.  If you need to differentiate I think you could have `source_no_emit` or `source_before_emit` but I'm not sure that is really needed.  The old name was fine I think.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 13]
```diff
@@ -94,9 +135,11 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
                               named_table_provider(table_names));
-        return DeclarationInfo{std::move(source_decl), num_columns};
+        return ProcessEmit(std::move(read),
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:140]: Same, revert to `source_decl`.

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 14]
```diff
@@ -199,12 +242,14 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
-      return DeclarationInfo{
-          compute::Declaration{
-              "scan", dataset::ScanNodeOptions{std::move(ds), std::move(scan_options)}},
-          num_columns};
+      DeclarationInfo no_emit_declaration = {
+          compute::Declaration{"scan", dataset::ScanNodeOptions{ds, scan_options}},
+          base_schema};
+
+      return ProcessEmit(std::move(read), std::move(no_emit_declaration),
+                         std::move(base_schema));
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:252]:
```suggestion
      DeclarationInfo scan_declaration = {
          compute::Declaration{"scan", dataset::ScanNodeOptions{ds, scan_options}},
          base_schema};

      return ProcessEmit(std::move(read), std::move(scan_declaration),
                         std::move(base_schema));
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 15]
```diff
@@ -222,19 +267,20 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
       }
       ARROW_ASSIGN_OR_RAISE(auto condition,
                             FromProto(filter.condition(), ext_set, conversion_options));
-
-      return DeclarationInfo{
+      DeclarationInfo no_emit_declaration{
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:270]:
```suggestion
      DeclarationInfo filter_declaration{
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 16]
```diff
@@ -222,19 +267,20 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
           compute::Declaration::Sequence({
               std::move(input.declaration),
               {"filter", compute::FilterNodeOptions{std::move(condition)}},
           }),
-          input.num_columns};
+          input.output_schema};
+
+      return ProcessEmit(std::move(filter), std::move(no_emit_declaration),
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:277]:
```suggestion
      return ProcessEmit(std::move(filter), std::move(filter_declaration),
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 17]
```diff
@@ -244,23 +290,48 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+            project_schema->AddField(
+                num_columns + static_cast<int>(project.expressions().size()) - 1,
+                std::move(project_field)));
+        i++;
+        expressions.emplace_back(des_expr);
       }
 
-      auto num_columns = static_cast<int>(expressions.size());
-      return DeclarationInfo{
+      DeclarationInfo no_emit_declaration{
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:326]:
```suggestion
      DeclarationInfo project_declaration{
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 18]
```diff
@@ -244,23 +290,48 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
           compute::Declaration::Sequence({
               std::move(input.declaration),
               {"project", compute::ProjectNodeOptions{std::move(expressions)}},
           }),
-          num_columns};
+          project_schema};
+
+      return ProcessEmit(std::move(project), std::move(no_emit_declaration),
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:333]:
```suggestion
      return ProcessEmit(std::move(project), std::move(project_declaration),
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 19]
```diff
@@ -338,15 +409,29 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+
+      // Create output schema from left, right relations and join keys
+      std::shared_ptr<Schema> join_schema = left.output_schema;
+      std::shared_ptr<Schema> right_schema = right.output_schema;
+
+      for (const auto& field : right_schema->fields()) {
+        ARROW_ASSIGN_OR_RAISE(
+            join_schema, join_schema->AddField(
+                             static_cast<int>(join_schema->fields().size()) - 1, field));
+      }
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:421]: Each time you call `AddField` it is going to create a new schema, create a new vector of fields, and copy all the fields over.  So this ends up having O(n^2) complexity.  Instead I think you could do something like...

```
FieldVector combined_fields = left.output_schema.fields();
const FieldVector& right_fields = right.output_schema.fields();
combined_fields.insert(combined_fields.end(), right_fields.begin(), right_fields.end());
std::shared_ptr<Schema> join_schema = schema(std::move(combined_fields));
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 20]
```diff
@@ -338,15 +409,29 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
       compute::Declaration join_dec{"hashjoin", std::move(join_options)};
-      auto num_columns = left.num_columns + right.num_columns;
       join_dec.inputs.emplace_back(std::move(left.declaration));
       join_dec.inputs.emplace_back(std::move(right.declaration));
-      return DeclarationInfo{std::move(join_dec), num_columns};
+
+      DeclarationInfo no_emit_declaration{std::move(join_dec), join_schema};
+
+      return ProcessEmit(std::move(join), std::move(no_emit_declaration),
+                         std::move(join_schema));
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:434]:
```suggestion
      DeclarationInfo join_declaration{std::move(join_dec), join_schema};

      return ProcessEmit(std::move(join), std::move(join_declaration),
                         std::move(join_schema));
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 21]
```diff
@@ -399,17 +495,38 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
+      }
+      // extract key fields to output schema
+      for (int id = 0; id < static_cast<int>(key_field_ids.size()); id++) {
+        output_fields.emplace_back(input_schema->field(key_field_ids[id]));
+      }
 
-      return DeclarationInfo{
+      std::shared_ptr<Schema> aggregate_schema = schema(std::move(output_fields));
+
+      DeclarationInfo no_emit_declaration{
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:522]:
```suggestion
      DeclarationInfo aggregate_declaration{
```

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 22]
```diff
@@ -399,17 +495,38 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
...
           compute::Declaration::Sequence(
               {std::move(input.declaration),
                {"aggregate", compute::AggregateNodeOptions{aggregates, keys}}}),
-          static_cast<int>(aggregates.size())};
+          aggregate_schema};
+
+      return ProcessEmit(std::move(aggregate), std::move(no_emit_declaration),
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/relation_internal.cc:528]:
```suggestion
      return ProcessEmit(std::move(aggregate), std::move(aggregate_declaration),
```

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 23]
```diff
@@ -121,6 +155,57 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
   return compute::Expression{std::move(modified_call)};
 }
 
+// TODO: complete this interface
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:158]: Does this still need completed?  Is there a follow-up JIRA?  What's the plan here?

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 24]
```diff
@@ -45,6 +54,31 @@ using internal::checked_cast;
 
 namespace engine {
 
+Status WriteParquetData(const std::string& path,
+                        const std::shared_ptr<fs::FileSystem> file_system,
+                        const std::shared_ptr<Table> input) {
+  EXPECT_OK_AND_ASSIGN(auto buffer_writer, file_system->OpenOutputStream(path));
+  PARQUET_THROW_NOT_OK(parquet::arrow::WriteTable(*input, arrow::default_memory_pool(),
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:61]: We are not in the `parquet::` namespace so it is not acceptable to throw.  Best to use `ARROW_RETURN_NOT_OK` instead or if you continue to use expect then use `EXPECT_OK`

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 25]
```diff
@@ -45,6 +54,31 @@ using internal::checked_cast;
 
 namespace engine {
 
+Status WriteParquetData(const std::string& path,
+                        const std::shared_ptr<fs::FileSystem> file_system,
+                        const std::shared_ptr<Table> input) {
+  EXPECT_OK_AND_ASSIGN(auto buffer_writer, file_system->OpenOutputStream(path));
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:60]: Generally a test method should either return a `Status` and not have any calls to `EXPECT_...` or it should return a value / void.

A good rule of thumb is:

Test case - Use ASSERT_, return void
Test helper method, used only within one file - Use ASSERT_, return void or use EXPECT_, return value
Test helper method, shared by multiple test files - Use ARROW_, return Status or Result

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 26]
```diff
@@ -121,6 +155,57 @@ inline compute::Expression UseBoringRefs(const compute::Expression& expr) {
...
+  ExtensionSet ext_set(ext_id_reg);
+  ASSERT_OK_AND_ASSIGN(auto sink_decls, DeserializePlans(
+                                            *buf, [] { return kNullConsumer; },
+                                            ext_id_reg, &ext_set, conversion_options));
+  auto other_declrs = sink_decls[0].inputs[0].get<compute::Declaration>();
+  arrow::AsyncGenerator<util::optional<compute::ExecBatch>> sink_gen;
+  auto sink_node_options = compute::SinkNodeOptions{&sink_gen};
+  auto sink_declaration = compute::Declaration({"sink", sink_node_options, "e"});
+  auto declarations = compute::Declaration::Sequence({*other_declrs, sink_declaration});
+  ASSERT_OK_AND_ASSIGN(auto acero_plan, compute::ExecPlan::Make(&exec_context));
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc:198]: You should move these lines into `GetTableFromPlan` (maybe change it to `GetTableFromDeclaration`)

**Review by westonpace — CHANGES_REQUESTED** [review]
A few naming suggestions and potential spots for cleanup but no complaints to the overall approach.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 2 above)
Comment (westonpace) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:518]: Another PR is fine but I wouldn't consider the output order from Acero to be too static.  Fixing it up to output things in the order Substrait expects would be nice so we can at least avoid the project node in some cases (when a direct emit).  It'll be a breaking change and probably cause some slight heartburn to our existing tests but we should probably fix it while we still have the opportunity.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 12 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:138]: Sure, we can stick to that.

File: cpp/src/arrow/engine/substrait/serde_test.cc (hunk 25 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:60]: Thank you for pointing this out. I will update accordingly.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 2 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:518]: Jira created here: https://issues.apache.org/jira/browse/ARROW-17656

File: cpp/src/arrow/engine/substrait/relation_internal.cc [hunk 27]
```diff
@@ -34,11 +34,53 @@ namespace arrow {
 namespace engine {
 
 template <typename RelMessage>
-Status CheckRelCommon(const RelMessage& rel) {
+bool HasEmit(const RelMessage& rel) {
+  if (rel.has_common()) {
+    return rel.common().has_emit();
```
[the remark this answers is not in the document]
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:47]: Ah my bad, I forgot to remove this. Not used now.

File: cpp/src/arrow/engine/substrait/relation_internal.cc (hunk 19 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/relation_internal.cc:438]: Thank you for catching this one.

File: cpp/src/arrow/engine/substrait/serde_test.cc (hunk 23 above)
Comment (vibhatha) [reply cpp/src/arrow/engine/substrait/serde_test.cc:197]: Removed this method and replaced with NamedTable. And this could be re-written later when we check file-system based things.

**Review by westonpace — APPROVED** [review]
Thanks again for putting this together.

**Comment by vibhatha** [comment]
Thank you for reviewing this one and keeping up with a few rounds of reviews.
